Step1

kimc1905 opened the pull request:
클래스를 분리했다가 다시합치는 작업을 몇번 반복했더니 늦었네요.
리뷰 부탁드립니다.

kimc1905 committed:
step1 구현

kimc1905 committed:
feat: Nonnull 어노테이션 추가

javajigi commented on src/main/java/bowling/domain/BowlingGame.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public boolean isEndGame() {
+        return currentFrame instanceof LastFrame && currentFrame.isEnd();
```
Frame에 isEndGame() 메소드를 구현하고 위임하면 좀 더 깔끔하게 구현 가능할 것 같아요.

javajigi commented on src/main/java/bowling/domain/BowlingGame.java:
```diff
@@ -0,0 +1,38 @@
+        this.frames = new LinkedList<>();
+        this.frames.add(new LastFrame());
+        IntStream.range(0, 9)
+                .forEach(i -> frames.addFirst(new Frame(frames.getFirst())));
```
`frames.addFirst(new Frame(frames.getFirst()))` 뭔가 좀 어색한데요.
다른 방식으로 구현할 방법은 없을까요?

javajigi commented on src/main/java/bowling/domain/DownPinCount.java:
```diff
@@ -0,0 +1,64 @@
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+public class DownPinCount {
```
이 클래스 추가 👍

javajigi commented on src/main/java/bowling/domain/DownPinCount.java:
```diff
@@ -0,0 +1,64 @@
+    public DownPinCount add(@Nonnull DownPinCount other) {
+        return new DownPinCount(count + other.count);
+    }
+
```
이런 메소드 추가하면 어떨까요?

```
public boolean isStrike() {
    return count == 10;
}

public boolean isSpare(DownPinCount secondThrow) {
    return count + secondThrow.count == 10;
}
```

kimc1905 replied:
좋은 지적 감사합니다. 
add에는 한 프레임에서 10개 이상의 핀이 쓰러지게끔 하는지 확인하는 의도가 있습니다.

의견 주신부분의 isSpare에서 아래와 같이 add를 사용하도록 변경하였습니다.
```java
public boolean isSpare(@Nonnull DownPinCount other) {
   return add(other).count == 10;
}
```

kimc1905 replied:
좋은 지적 감사합니다. 
add에는 한 프레임에서 10개 이상의 핀이 쓰러지게끔 하는지 확인하는 의도가 있습니다.

의견 주신부분의 isSpare에서 아래와 같이 add를 사용하도록 변경하였습니다.
```java
public boolean isSpare(@Nonnull DownPinCount other) {
   return add(other).count == 10;
}
```

javajigi reviewed (changes requested):
객체 추출 및 전체적인 구조가 정말 깔끔하네요. 제가 피드백할 부분이 거의 없네요. 그래도 의구심 나는 부분과 의견 몇 개 남겨 봤어요.

kimc1905 committed:
refactor(Frame, LastFrame): BowlingGame의 isEndGame 로직을 Frame 클래스에 위임

kimc1905 committed:
DownPinCount, Frame, LastFrame 리팩토링
refactor(DownPinCount): Spare 확인 메소드 추가, isAllDown 메소드 이름 변경
refactor(Frame, LastFrame): LastFrame에 있는 기능을 Frame으로 이동, 파라미터에 사용되는 Optional 제거
[Why should Java 8's Optional not be used in arguments]
(https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments/31923105)

kimc1905 committed:
Frame, LastFrame 오류 수정, 테스트 추가
refactor(Frame, LastFrame): `isEnd` 메소드 이름을 `isEndFrame`으로 변경
fix(Frame, LastFrame): Spare 확인 메소드에서 스트라이크인지 먼저 확인하도록 변경
fix(Frame): `add`를 할때 10개 이상의 핀이 쓰러지는 확인할수 있도록 `isEndFrame` 기능 변경
test(DownPinCountTest): 스페어 확인 테스트 추가

kimc1905 committed:
`BowlingGame`에서 `Frame`을 만드는 부분 가독성 높이기
refactor(BowlingGame): 가독성을 높이기 위해 지역변수 `nextFrame`을 명시적으로 둠.

kimc1905 commented:
피드백 주신 내용은 반영 하였습니다. 
추가적으로 이번 구현에서 Optional을 적극적으로 썼는데 좀 과했던것 같네요.
커밋 메시지에도 남겼지만 파라미터에 사용되는 Optional은 제거 하였습니다.
